### ISSUE TITLE
Connection Reset with FutureResponse and body-limit

### ISSUE THREAD

**Issue by kolbma** [actix/actix-web#563]
Hi,

I think that there might be some bug in handling the body limit when using `req.body()` futures.

My tests with the `TestServer::with_factory` doing POST requests, which hit the body limit, I sometimes/often but not always get connection resets...

```
ParseError(Error(Io(Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." })))', libcore\result.rs:945:5
```

All the other tests staying below the limit don't have any problems.  
There is no difference if the tests run with `cargo test -- --test-threads=1` or without thread limitation.

Because I am pretty new to Rust and Actix, I'm not sure what is the best to do, to get the cause.
Do you have some instructions?!

**Comment by kolbma** [comment]
I've extracted a test and ran it against an external server process. So I could capture the net traffic...

![capture](https://user-images.githubusercontent.com/5228369/47531282-6c236180-d8ad-11e8-89bd-74068dc9db93.PNG)

The server part is behaving correctly I think.  
A problem seems to be in the **test server/client** part.  
It seems that it doesn't handle the received (has it received? at least the server sent it) 413 response when the socket connection drops too fast.  
I think the test client should react to the **413** (or any other response), although it is still sending data to the server, shouldn't it?

**Comment by kolbma** [comment]
Does it do, what it says (block)?

https://github.com/actix/actix-web/blob/42d5d48e7105270e89eb8af6b111c305c5536e00/src/test.rs#L208

**Comment by DoumanAsh** [comment]
@kolbma yes, `sefl.rt.block_on` waits for future to be finished

As for body limit, I'm not sure right now, but I think the server part should only return error response and close connection
I think connection resets are normal in this case?
But I'm not sure do you get it on client or what?

**Comment by kolbma** [comment]
Sometimes the client handles the response and sometimes it detects socket close and gets panic. Seems to be a race cond.

**Comment by rofrol** [comment]
@kolbma Any news on that?

**Comment by fafhrd91** [comment]
@rofrol did you try master?

**Comment by rofrol** [comment]
@fafhrd91 I haven't try to reproduce that, but was curious about the status. Do we have some tests that do this?

**Comment by fafhrd91** [comment]
There is no test for that, but pr is welcomed

**Comment by fafhrd91** [comment]
should be fixed in 0.7.19
